### PR TITLE
Redo function wrappers implementation.

### PR DESCRIPTION
**Pull request by lcrh**
This changes the implementation of function wrappers. Function wrappers used to be handled in two different ways: Instance wrappers for non-bound functions (i.e., non-methods) and another more traditional approach for member functions. This PR removes this in favor of common handling, somewhat inspired by wrapt.

FunctionWrappers are now directly supported by a resource_registry as special wrapper types. Each FunctionWrapper type is associated with a MethodWrapper type which can be used in cases where the the wrapped object is not a function, but a bound method. A pair of FunctionWrapper and MethodWrapper types is automatically created when registering a function.

The new wrapping process has the following levels:
1) Wrapt replaces the original python function with a wrapt function proxy that routes calls into a FunctionWrapper or MethodWrapper invokable, depending on whether the function is called unbound or as a bound method.
2) These invokables call the the original, wrapped function internally. MethodWrapper also stores the instance.
2) The wrapt function proxy is then registered against the FunctionWrapper type, that is, the function wrapper can be committed and checked out of a store.

To support this, FunctionWrapper and MethodWrapper have an API that exposes both the wrapt proxy *and* the function wrapped by the proxy.

This also adds support for registering async functions and methods.

### CODE REVIEW

File: src/enact/function_wrappers.py [hunk 1]
```diff
@@ -192,28 +299,113 @@ def create(cls, function_name: str, fun: Callable) -> 'Call':
...
         'Cannot repeat a call of an unregistered function. '
         'Please make sure your top-level function is registered.')
     args, kwargs = resource.to_python_args()
-    result = self._callable(*args, **kwargs)
-    return resource_registry.wrap(result)
+    return self._callable(*args, **kwargs)
+
+@resource_registry.register
+@dataclasses.dataclass
+class AsyncCall(invocations.AsyncInvokable[CallArgs, Any]):
```
Comment (isaacery) [inline src/enact/function_wrappers.py:313]: I guess there's no concept of a private constructor in Python? That would be the classic way to implement this factory pattern kinda thing. It seems strange that you can initialize an AsyncCall without a callable, but not sure there is a way around it.

**Review by isaacery — APPROVED** [review]
Generally looks good, see my comments for some questions/nits.

File: src/enact/resource_registry.py [hunk 2]
```diff
@@ -186,6 +190,82 @@ def set_wrapped_value(cls, target: dict, src: dict):
...
+    """The python function that is being called by this invokable."""
+    raise NotImplementedError()
+
+  @classmethod
+  def wrap(cls: Type[FunctionWrapperT], c: Callable) -> Union[
+      'FunctionWrapper', 'MethodWrapper']:
+    """Wrap a function."""
+    if inspect.ismethod(c):
+      return cls.method_wrapper().wrap(c)
+    assert c == cls.wrapper_function(), (
```
[the remark this answers is not in the document]
Comment (lcrh) [reply src/enact/resource_registry.py:229]: The wrapper function replaces the wrapped function in python land, so this is correct.

File: src/enact/resource_registry.py [hunk 3]
```diff
@@ -186,6 +190,82 @@ def set_wrapped_value(cls, target: dict, src: dict):
...
+  def wrapper_function(cls) -> Callable:
+    """The python function that routes calls into this invokable."""
+    raise NotImplementedError()
+
+  @classmethod
+  def wrapped_function(cls) -> Callable:
+    """The python function that is being called by this invokable."""
+    raise NotImplementedError()
+
+  def get_instance(self):
```
[the remark this answers is not in the document]
Comment (lcrh) [reply src/enact/resource_registry.py:263]: Done.

File: src/enact/function_wrappers.py [hunk 4]
```diff
@@ -30,150 +29,258 @@
...
 @dataclasses.dataclass
 class CallArgs(resources.Resource):
-  """Wrapper for call arguments."""
-  args: List[interfaces.FieldValue]
-  kwargs: Dict[str, interfaces.FieldValue]
+  """Represents call arguments to a python function.
+
+  This class is used for wrapping python functions.
+  """
+  args: List
```
[the remark this answers is not in the document]
Comment (lcrh) [reply src/enact/function_wrappers.py:42]: I prefer `List` to `List[Any]`

File: src/enact/function_wrappers.py (hunk 1 above)
Comment (lcrh) [reply src/enact/function_wrappers.py:313]: Not that I'm aware of.

**Pull request merged**